src/skagent/model.py

```diff
@@ -165,46 +318,116 @@ def simulate_dynamics(
+            if agent_count == 1:
+                agent_count = val.size
+            elif agent_count != val.size:
+                raise ValueError(
+                    f"Inconsistent agent counts in dynamics: variable '{var}' has size {val.size}, "
+                    f"but expected size {agent_count}"
+                )
+
     for sym in dynamics:
         # Using the fact that Python dictionaries are ordered
         feq = dynamics[sym]
 
         if isinstance(feq, Control):
-            # This tests if the decision rule is age varying.
-            # If it is, this will be a vector with the decision rule for each agent.
+            # Initialize control array for multiple agents
+            if agent_count > 1:
+                vals[sym] = np.zeros(agent_count)
+
+            # This tests if the decision rule is age varying or agent-heterogeneous
             if isinstance(dr[sym], np.ndarray):
-                ## Now we have to loop through each agent, and apply the decision rule.
-                ## This is quite slow.
+                # Now we have to loop through each agent, and apply the decision rule.
                 for i in range(dr[sym].size):
                     vals_i = {
                         var: (
                             vals[var][i]
-                            if isinstance(vals[var], np.ndarray)
-                            else vals[var]
+                            if isinstance(vals[var], np.ndarray) and vals[var].size > 1
+                            else (
+                                vals[var].item()
+                                if isinstance(vals[var], np.ndarray)
+                                and vals[var].size == 1
+                                else vals[var]
+                            )
                         )
                         for var in vals
                     }
                     vals[sym][i] = dr[sym][i](
                         *[vals_i[var] for var in signature(dr[sym][i]).parameters]
                     )
             else:
-                vals[sym] = dr[sym](
-                    *[vals[var] for var in signature(dr[sym]).parameters]
-                )  # TODO: test for signature match with Control
+                # Single decision rule applied to all agents
+                if agent_count > 1:
+                    # Apply decision rule agent by agent
+                    for i in range(agent_count):
+                        vals_i = {
+                            var: (
+                                vals[var][i]
+                                if isinstance(vals[var], np.ndarray)
+                                and vals[var].size > 1
+                                else (
+                                    vals[var].item()
+                                    if isinstance(vals[var], np.ndarray)
+                                    and vals[var].size == 1
+                                    else vals[var]
```

Comment:
This support for age-varying decision rules isn't the same thing as the ex ante heterogeneity feature.

Would it be possible to split that out into a separate PR?

I have some resistance to it, and I don't want it to get in the way of the ex ante heterogeneous calibration and simulation stuff.